### PR TITLE
image-tagging-feat

### PR DESCRIPTION
**Pull request by gireesh-devtron**

### CODE REVIEW

File: src/Common/Common.service.ts [hunk 1]
```diff
@@ -20,4 +20,19 @@ export const getTeamListMin = (): Promise<TeamList> => {
           result: list,
       };
   });
-};
\ No newline at end of file
+};
+
+interface UserRole extends ResponseType {
```
Comment (vivek-devtron) [inline src/Common/Common.service.ts:25]: Please move this to types.

File: yarn.lock
Comment (vivek-devtron) [inline yarn.lock:1]: Please remove this file

File: src/Common/Types.ts [hunk 2]
```diff
@@ -296,33 +326,57 @@ export interface MaterialInfo {
...
     materials: any[]
     userApprovalConfig: UserApprovalConfigType
     requestedUserId: number
-}
\ No newline at end of file
+    tagsEditable: boolean
+    appReleaseTagNames: string[]
+}
+
+export interface ReleaseTag {
```
Comment (vivek-devtron) [inline src/Common/Types.ts:338]: we should move release tag related types to separate type file

File: src/Common/ImageTags.tsx [hunk 3]
```diff
@@ -0,0 +1,594 @@
...
+    }
+
+    const handleTagSoftDelete = (index) => {
+        const updatedTags = [...displayedTags]
+        updatedTags[index] = {
+            ...updatedTags[index],
+            deleted: !updatedTags[index].deleted,
+        }
+        const updatedTag = updatedTags[index]
+        let updatedSoftDeleteTags = [...softDeleteTags]
```
Comment (vivek-devtron) [inline src/Common/ImageTags.tsx:166]: Please change it to const

File: src/Common/ImageTags.tsx [hunk 4]
```diff
@@ -0,0 +1,594 @@
...
+                            <div className="form__error">{tagErrorMessage}</div>
+                        </div>
+                    )}
+                    <div className="flex-wrap mt-8 flex left">
+                        {displayedTags?.map((tag, index) => (
+                            <ImageTagButton
+                                key={tag.id}
+                                text={tag?.tagName}
+                                isSoftDeleted={tag?.deleted}
+                                isEditing={isEditing}
```
Comment (vivek-devtron) [inline src/Common/ImageTags.tsx:396]: Please avoid inline function wherever possible

File: src/Common/ImageTags.tsx [hunk 5]
```diff
@@ -0,0 +1,594 @@
...
+    const [showTagsWarning, setShowTagsWarning] = useState(false)
+    const [displayedTags, setDisplayedTags] = useState<ReleaseTag[]>(imageReleaseTags ? imageReleaseTags : [])
+    const [tagErrorMessage, setTagErrorMessage] = useState('')
+    const [createTags, setCreateTags] = useState<ReleaseTag[]>([])
+    const [softDeleteTags, setSoftDeleteTags] = useState<ReleaseTag[]>([])
+    const [hardDeleteTags, setHardDeleteTags] = useState<ReleaseTag[]>([])
+    const [isSuperAdmin, setSuperAdmin] = useState<boolean>(false)
+    const [descriptionValidationMessage, setDescriptionValidationMessage] = useState<string>('')
+    const [textInput, setTextInput] = useState<string>('')
+    const [saveData, setSaveData] = useState<boolean>(false)
```
Comment (vivek-devtron) [inline src/Common/ImageTags.tsx:50]: We can remove this variable. can be handled without this

File: src/Common/ImageTags.tsx [hunk 6]
```diff
@@ -0,0 +1,594 @@
...
+            <TippyCustomized
+                theme={TippyTheme.white}
+                className="w-300 fcv-5"
+                placement="right"
+                Icon={QuestionFilled}
+                heading="Release tags"
+                showCloseButton={true}
+                trigger="click"
+                interactive={true}
+                documentationLinkText="View Documentation"
```
Comment (vivek-devtron) [inline src/Common/ImageTags.tsx:283]: Where is the doc link?

File: src/Common/ImageTags.tsx [hunk 7]
```diff
@@ -0,0 +1,594 @@
...
+                additionalContent={getBuildContextAdditionalContent()}
+            >
+                <div className="icon-dim-16 fcn-5 ml-8 cursor">
+                    <Question />
+                </div>
+            </TippyCustomized>
+        )
+    }
+
+    const getBuildContextAdditionalContent = () => {
```
Comment (vivek-devtron) [inline src/Common/ImageTags.tsx:293]: rename this method

File: src/Common/ImageTags.tsx [hunk 8]
```diff
@@ -0,0 +1,594 @@
...
+
+    const creatableRef = useRef(null)
+
+    if (newDescription === '' && displayedTags.length === 0 && !isEditing) {
+        return tagsEditable ? (
+            <div className="bcn-0 mt-8 add-image-button">
+                <AddImageButton handleEditClick={handleEditClick} />
+            </div>
+        ) : (
+            <div></div>
```
Comment (vivek-devtron) [inline src/Common/ImageTags.tsx:321]: We can add self closing div

File: src/Common/ImageTags.tsx [hunk 9]
```diff
@@ -0,0 +1,594 @@
...
+    }
+
+    const handleKeydown = (event) => {
+        if (event.key === 'Enter' && event.target.value.trim()) {
+            handleTagCreate(event.target.value.trim())
+        }
+    }
+
+    const setInputValue = (e) => {
+        e.target.value && validateTag(e.target.value)
```
Comment (vivek-devtron) [inline src/Common/ImageTags.tsx:332]: Why are we doing the validation on every keyStroke?

Comment (ShashwatDadhich) [reply src/Common/ImageTags.tsx:332]: It is said by the product to validate on every stroke

File: src/Common/ImageTags.tsx [hunk 10]
```diff
@@ -0,0 +1,594 @@
...
+    const handleSave = async () => {
+        if (tagErrorMessage) return
+
+        const payload = {
+            createTags: createTags,
+            softDeleteTags: softDeleteTags,
+            imageComment: {
+                id: 0,
+                comment: newDescription.trim(),
+                artifactId: 0,
```
[the remark this answers is not in the document]
Comment (gireesh-devtron) [reply src/Common/ImageTags.tsx:219]: this is not required in the payload

File: src/Common/ImageTags.tsx (hunk 5 above)
Comment (gireesh-devtron) [reply src/Common/ImageTags.tsx]: done

File: src/Common/ImageTags.tsx (hunk 6 above)
Comment (gireesh-devtron) [reply src/Common/ImageTags.tsx:274]: there is no link,everythink is mention in the additional content

**Review by vivek-devtron — APPROVED** [review]

File: src/Common/Types.ts (hunk 2 above)
Comment (gireesh-devtron) [reply src/Common/Types.ts]: done

**Pull request merged**